File: app/src/main/java/com/rosebay/odds/ui/favoriteOdds/FavoriteOddsPresenterImpl.java
package com.rosebay.odds.ui.favoriteOdds;

import android.annotation.SuppressLint;
import android.support.annotation.VisibleForTesting;

import com.rosebay.odds.OddsApplication;
import com.rosebay.odds.localStorage.FavoriteDao;
import com.rosebay.odds.model.Favorite;
import com.rosebay.odds.model.SingleOdd;
import com.rosebay.odds.network.FirebaseClient;

import java.util.ArrayList;
import java.util.List;

import javax.annotation.Nullable;
import javax.inject.Inject;

import easymvp.AbstractPresenter;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class FavoriteOddsPresenterImpl extends AbstractPresenter<FavoriteOddsView> implements FavoriteOddsPresenter {

    @Inject
    FavoriteDao favoriteDao;

    @Inject
    FirebaseClient firebaseClient;

    @VisibleForTesting
    FavoriteOddsView favoriteOddsView;

    List<SingleOdd> mSingleOddsList = new ArrayList<>();

    @SuppressLint("CheckResult")
    public void getAllFavorites() {
        Observable.fromCallable(() -> favoriteDao.getUserFavorites())
            .flatMapIterable(items -> {
                if (items != null) {
                    return items;
                } else {
                    throw new NullPointerException();
                }
            })
            .map(Favorite::getPostId)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(result -> fetchFromFirebase(result),
                    e -> getView().noFavorites());
        }

    public void fetchFromFirebase(String post) {
        getView().onLoading();
        Disposable disposable = firebaseClient.fetchSingleOdd(post)
            .map(item -> mSingleOddsList.add(item))
            .doOnError(e -> getView().onError())
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(result -> getView().setData(mSingleOddsList));
    }

    @Override
    public void onViewAttached(FavoriteOddsView view) {
        super.onViewAttached(view);
        OddsApplication.getAppComponent().inject(this);
    }

    @Override
    public void onViewDetached() {
        super.onViewDetached();
    }

    @Nullable
    @Override
    public FavoriteOddsView getView() {
        return super.getView() == null ? favoriteOddsView : super.getView();
    }
}

File: app/src/main/java/com/rosebay/odds/ui/favoriteOdds/FavoriteOddsPresenterImpl.kt
package com.rosebay.odds.ui.favoriteOdds

import androidx.annotation.VisibleForTesting
import com.rosebay.odds.OddsApplication
import com.rosebay.odds.localStorage.FavoriteDao
import com.rosebay.odds.model.Favorite
import com.rosebay.odds.network.FirebaseClient
import com.rosebay.odds.util.EmptyResponseException
import easymvp.AbstractPresenter
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.exceptions.CompositeException
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

open class FavoriteOddsPresenterImpl @Inject constructor() : AbstractPresenter<FavoriteOddsView>(), FavoriteOddsPresenter {

    @Inject
    lateinit var favoriteDao: FavoriteDao

    @Inject
    lateinit var firebaseClient: FirebaseClient

    @VisibleForTesting
    var favoriteOddsView: FavoriteOddsView? = null

    override fun getAllFavorites() {
        view?.onLoading()
        Observable.fromCallable { favoriteDao.userFavorites }
                .map { it ->
                    if (it.isEmpty()) {
                        throw EmptyResponseException()
                    } else {
                        it
                    }
                }
                .flatMapIterable<Favorite> { items -> items}
                .map { it -> firebaseClient.fetchSingleOdd(it.postId) }
                .flatMap { item -> item.toObservable() }
                .toList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe( { result -> view?.setData(result) },
                    { e ->
                        when (e) {
                            is EmptyResponseException -> {
                                view?.noFavorites()
                            }
                            is CompositeException -> for (exception in e.exceptions) {
                                if (exception is EmptyResponseException) {
                                    view?.noFavorites()
                                }
                            }
                            else ->  {
                                view?.onError()
                            }
                        }
                    })
    }

    override fun onViewAttached(view: FavoriteOddsView) {
            super.onViewAttached(view)
            OddsApplication.appComponent.inject(this)
        }

    override fun onViewDetached() {
            super.onViewDetached()
        }

    override fun getView(): FavoriteOddsView? {
        return if (super.getView() == null) favoriteOddsView else super.getView()
    }
}
